File: questao3.py
import numpy as np
import cv2

def bgrFilter(src, bgr):
	for y in range(len(src)):
		for x in range(len(src[y])):
			if (src[y, x, 0] == bgr[0] and src[y, x, 1] == bgr[1] and src[y, x, 2] == bgr[2]) == False:
				src[y, x, 0] = 255
				src[y, x, 1] = 255
				src[y, x, 2] = 255
				
colors = {0 : [255, 0, 255], 1 : [255, 0, 0], 2 : [255, 255, 0], 3 : [0, 0, 0], 4 : [0, 255, 0], 5 : [0, 255, 255], 6 : [0, 0, 255]}

order = [0, 1, 2, 3, 4, 5, 6]

for i in order:
	color = colors[i]

	img = cv2.imread("targets.png")
	
	for i in range(100):
		kernel = np.ones((5, 5), np.float32) / 25
		img = cv2.filter2D(img, -1, kernel)

	bgrFilter(img, color)
	print("color=", color)
	
	gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
	gray = cv2.medianBlur(gray, 5)
	
	circles = cv2.HoughCircles(gray, cv2.HOUGH_GRADIENT, 1, 20, param1 = 50, param2 = 30, minRadius = 0, maxRadius = 0)
	#circles = np.uint16(np.around(circles))
	
	if circles is not None:
		circle = circles[0, 0, :]
		print("x=%d, y=%d, radius=%d" % (circle[0], circle[1], circle[2]))

File: questao3.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <map>

using namespace cv;
using namespace std;

Vec3b darkBlue(255, 0, 0);
Vec3b blue(255, 255, 0);
Vec3b black(159, 159, 159);
Vec3b green(0, 255, 0);
Vec3b yellow(0, 255, 255);
Vec3b red(0, 0, 255);
Vec3b magenta(255, 0, 255);

map<int, Vec3b> colors = {{0, darkBlue}, {1, blue}, {2, black}, {3, green}, {4, yellow}, {5, red}, {6, magenta}};

vector<Point> getTargets(int order[]) {
	unsigned int numTargets = 7;

	Mat img, kernel, gray;
	img = imread("targets.png");

	kernel = Mat::ones(5, 5, CV_64F) / 25;

	filter2D(img, img, -1, kernel);

	cvtColor(img, gray, CV_BGR2GRAY);
	GaussianBlur(gray, gray, Size(9, 9), 2, 2);
	vector<Vec3f> circles;
	HoughCircles(gray, circles, CV_HOUGH_GRADIENT, 1, gray.rows / 8, 200, 100, 0, 0);

	vector<Point> targets;

	unsigned int i, j;
	for (i = 0; i < numTargets; i++) {
		Vec3b color = colors[order[i]];
		for (j = 0; j < circles.size(); j++) {
			Vec3b intensity = img.at<Vec3b>(circles[j][1], circles[j][0]);
			if (intensity == color)
				targets.push_back(Point(circles[j][0], circles[j][1]));
		}
	}

	return targets;
}

int main() {
	int order[] = {0, 1, 2, 3, 4, 5, 6};

	vector<Point> targets = getTargets(order);

	for (size_t i = 0; i < targets.size(); i++)
		cout << "x=" << targets[i].x << ", y=" << targets[i].y << endl;

	return 0;
}